aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -254,6 +254,157 @@ Tensor matrix_rank(const Tensor& self, bool symmetric) {
+    const std::vector<std::vector<int64_t>>& order,
+    int64_t i,
+    int64_t j) {
+  if (i == j) {
+    return tensors[i];
+  }
+  return at::mm(
+      matrix_chain_multiplication(tensors, order, i, order[i][j]),
+      matrix_chain_multiplication(tensors, order, order[i][j] + 1, j));
+}
+
+} // namespace
+
+Tensor linalg_multi_dot(TensorList _tensors) {
+  const auto n = _tensors.size();
+  TORCH_CHECK(n >= 2, "multi_dot() expected at least 2 tensors but got ", n);
+
+  std::vector<int64_t> out_shape;
+  std::vector<Tensor> tensors(n);
+
+  // If the first tensor is 1D view it as a row vector
+  if (_tensors[0].dim() == 1) {
+    tensors[0] = _tensors[0].view({1, -1});
+  } else if (_tensors[0].dim() == 2) {
+    tensors[0] = _tensors[0];
+    out_shape.push_back(tensors[0].size(0));
+  } else {
+    TORCH_CHECK(
+        false,
+        "multi_dot() the first tensor must be 1D or 2D but got ",
+        _tensors[0].dim());
+  }
+
+  // If the last tensor is 1D view it as a column vector
+  if (_tensors[n - 1].dim() == 1) {
+    tensors[n - 1] = _tensors[n - 1].view({-1, 1});
+  } else if (_tensors[n - 1].dim() == 2) {
+    tensors[n - 1] = _tensors[n - 1];
+    out_shape.push_back(tensors[n - 1].size(1));
+  } else {
+    TORCH_CHECK(
+        false,
+        "multi_dot() the last tensor must be 1D or 2D but got ",
+        _tensors[0].dim());
+  }
+
+  // Ensure middle tensors are 2D
+  for (auto i = decltype(n){1}; i < n - 1; ++i) {
+    TORCH_CHECK(
+        _tensors[i].dim() == 2,
+        "multi_dot() tensor ",
+        i,
+        " must be 2D but got ",
+        _tensors[0].dim());
+    tensors[i] = _tensors[i];
+  }
+
+  // Ensure all tensors are on the same device and have the same dtype
+  const auto dtype = tensors[0].dtype();
+  const auto device = tensors[0].device();
+  bool has_empty_tensor = tensors[0].numel() == 0;
+  for (auto i = decltype(n){1}; i < n; ++i) {
+    TORCH_CHECK(
+        tensors[i].dtype() == dtype,
+        "multi_dot() all tensors have be the same dtype but tensor 0 is ",
+        dtype,
+        " and tensor ",
+        i,
+        " ",
+        tensors[i].dtype());
+    TORCH_CHECK(
+        tensors[i].device() == device,
+        "multi_dot() all tensors must be on the same device but tensor 0 is on ",
+        device,
+        " and tensor ",
+        i,
+        " on ",
+        tensors[i].device());
+     has_empty_tensor |= tensors[i].numel() == 0;
+  }
+
+  if (has_empty_tensor) {
+    return at::zeros(out_shape, tensors[0].options());
+  }
+
+  if (tensors.size() == 2) {
+    return at::mm(tensors[0], tensors[1]).view(out_shape);
+  }
+
+  const auto order = matrix_chain_order(tensors);
+  Tensor result = matrix_chain_multiplication(tensors, order, 0, n - 1);
+  return result.view(out_shape);
+}
+
+Tensor& linalg_multi_dot_out(TensorList tensors, Tensor& result) {
+  Tensor res = at::linalg_multi_dot(tensors);
+  at::native::resize_output(result, res.sizes());
+  result.copy_(res);
```

Comment:
This copy is going to add another linear algebra operator with wrong out= behavior we need to fix later. And this is the release where we want to fix those issues!

Can this algorithm be refactored to use mm_out for the final mm?

Reply:
Yes, I'll refactor it.